chain/network/src/routing.rs

```diff
@@ -758,100 +759,180 @@ pub struct RoutingTableInfo {
+    p2id: HashMap<PeerId, usize>,
+    id2p: FxHashMap<usize, PeerId>,
+    adjacency: FxHashMap<usize, FxHashSet<usize>>,
     total_active_edges: u64,
+    id2counter: FxHashMap<usize, i32>,
+    unused: FxHashSet<usize>,
 }
 
 impl Graph {
     pub fn new(source: PeerId) -> Self {
-        Self { source, adjacency: HashMap::new(), total_active_edges: 0 }
+        let mut res = Self {
+            source: source.clone(),
+            p2id: HashMap::default(),
+            id2p: FxHashMap::default(),
+            adjacency: FxHashMap::default(),
+            total_active_edges: 0,
+            id2counter: FxHashMap::default(),
+            unused: FxHashSet::default(),
+        };
+        res.id2p.insert(0, source.clone());
+        res.p2id.insert(source, 0);
+        res.change_counter(0, 1);
+
+        res
     }
 
-    fn contains_edge(&mut self, peer0: &PeerId, peer1: &PeerId) -> bool {
-        if let Some(adj) = self.adjacency.get(&peer0) {
-            if adj.contains(&peer1) {
-                return true;
+    fn contains_edge(&self, peer0: &PeerId, peer1: &PeerId) -> bool {
+        if let Some(id0) = self.p2id.get(&peer0) {
+            if let Some(id1) = self.p2id.get(&peer1) {
+                if let Some(adj) = self.adjacency.get(id0) {
+                    if adj.contains(&id1) {
+                        return true;
+                    }
+                }
             }
         }
-
         false
     }
 
-    fn add_directed_edge(&mut self, peer0: PeerId, peer1: PeerId) {
-        self.adjacency.entry(peer0).or_insert_with(HashSet::new).insert(peer1);
+    fn change_counter(&mut self, id: usize, counter: i32) {
+        let val = self.id2counter.entry(id).or_default();
+        *val += counter;
+
+        if *val == 0 && id != 0 {
+            let peer = self.id2p.get(&id).take().cloned().unwrap();
+            self.p2id.remove(&peer);
+            self.unused.insert(id);
+        }
+    }
+
+    fn get_id(&mut self, peer0: &PeerId) -> usize {
+        match self.p2id.entry(peer0.clone()) {
+            Entry::Occupied(occupied) => *occupied.get(),
+            Entry::Vacant(vacant) => {
+                let val =
+                    if let Some(val) = self.unused.drain().next() { val } else { self.id2p.len() };
+
+                vacant.insert(val);
+                self.id2p.insert(val, peer0.clone());
+                val
+            }
+        }
     }
 
-    fn remove_directed_edge(&mut self, peer0: &PeerId, peer1: &PeerId) {
-        self.adjacency.get_mut(&peer0).unwrap().remove(&peer1);
+    fn add_edge_internal(&mut self, peer0: PeerId, peer1: PeerId) {
+        if peer0 != peer1 {
+            let id0 = self.get_id(&peer0);
+            let id1 = self.get_id(&peer1);
+
+            self.adjacency.entry(id0).or_default().insert(id1);
+            self.adjacency.entry(id1).or_default().insert(id0);
+
+            self.change_counter(id0, 1);
+            self.change_counter(id1, 1);
```

Comment:
Is the graph undirected? If that's the case, I thing we don't need to maintain counts, as the count seems to always be equal to `self.adjecency[id].len()`? Can we use `adjecency[id].is_empty()` as a gc condition(/

Reply:
Thanks, yes it's undirected. I simplified the code.